crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+ * @param plain is an output
+ * @param plainlen is the length of plain
+ * @param cipher is an output
+ * @param cipherlen input/output, better be big enough on input, exact on output
+ * @return 1 for good otherwise bad
+ */
+static int hpke_aead_enc(OSSL_LIB_CTX *libctx,
+                         OSSL_HPKE_SUITE suite,
+                         unsigned char *key, size_t keylen,
+                         unsigned char *iv, size_t ivlen,
+                         unsigned char *aad, size_t aadlen,
+                         unsigned char *plain, size_t plainlen,
+                         unsigned char *cipher, size_t *cipherlen)
+{
+    int erv = 1;
+    EVP_CIPHER_CTX *ctx = NULL;
+    int len;
+    size_t ciphertextlen;
+    unsigned char *ciphertext = NULL;
+    size_t taglen = 0;
+    uint16_t aead_ind = 0;
+    EVP_CIPHER *enc = NULL;
+    unsigned char tag[16];
+
+    aead_ind = aead_iana2index(suite.aead_id);
+    if (aead_ind == 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    taglen = hpke_aead_tab[aead_ind].taglen;
+    if (taglen != 16) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if ((taglen + plainlen) > *cipherlen) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /*
+     * Allocate this much extra for ciphertext and check the AEAD
+     * doesn't require more - If it does, we'll fail.
+     */
+    ciphertext = OPENSSL_malloc(plainlen + taglen);
+    if (ciphertext == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Create and initialise the context */
+    if (!(ctx = EVP_CIPHER_CTX_new())) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Initialise the encryption operation. */
+    enc = EVP_CIPHER_fetch(libctx, hpke_aead_tab[aead_ind].name, NULL);
+    if (enc == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_EncryptInit_ex(ctx, enc, NULL, NULL, NULL) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    EVP_CIPHER_free(enc);
+    enc = NULL;
+    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, ivlen, NULL) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Initialise key and IV */
+    if (EVP_EncryptInit_ex(ctx, NULL, NULL, key, iv) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /*
+     * Provide any AAD data. This can be called zero or more times as
+     * required
+     */
+    if (aadlen != 0 && aad != NULL) {
+        if (EVP_EncryptUpdate(ctx, NULL, &len, aad, aadlen) != 1) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+    }
+    /*
+     * Provide the message to be encrypted, and obtain the encrypted output.
+     * EVP_EncryptUpdate can be called multiple times if necessary
+     */
+    if (EVP_EncryptUpdate(ctx, ciphertext, &len, plain, plainlen) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    ciphertextlen = len;
+    /*
+     * Finalise the encryption. Normally ciphertext bytes may be written at
+     * this stage, but this does not occur in GCM mode
```

Comment:
We may not be using GCM? Although it doesn' t occur for chacha20-poly1305 either.

Reply:
fair enough, fixed